torch/csrc/jit/codegen/cuda/maxinfo_propagator.cpp

```diff
@@ -28,88 +36,114 @@ void MaxInfoPropagator::run() {
   // std::list instead of std::priority_queue because C++'s
   // std::priority_queue does not support increase-key, and might not be
   // deterministic either.
-  std::list<NextHopInfo> propagation(1);
-  propagation.back().from = nullptr;
-  propagation.back().to = reference;
-  propagation.back().info_to = reference_info;
+  std::list<NextHopWithInfo> candidates(1);
+  candidates.back().next_hop.from = nullptr;
+  candidates.back().next_hop.to = reference_;
+  candidates.back().info_to = reference_info_;
 
-  // Insert the given next hop the correct position in `propagation`. If there
+  // Insert the given next hop the correct position in `candidates`. If there
   // is an existing next hop that preserves more information, then we will just
   // discard `info`.
-  auto insertNextHopInfo = [&](const NextHopInfo& info) {
+  auto insertNextHop = [&](const NextHopWithInfo& info) {
     if (!*(info.info_from)) {
       // When there is no more information about the starting tensor,
-      // we are not interested in continuing the propagation.
+      // we are not interested in continuing the path-finding.
       return;
     }
     // Find if there is already a path to the dest tensor
     auto existing = std::find_if(
-        propagation.begin(), propagation.end(), [&](const NextHopInfo& i) {
-          return i.to == info.to;
+        candidates.begin(), candidates.end(), [&](const NextHopWithInfo& i) {
+          return i.next_hop.to == info.next_hop.to;
         });
     // Only insert if there is no existing path to the dest tensor, or the new
     // path preserves more information about the starting tensor.
-    if (existing == propagation.end() || *existing < info) {
-      if (existing != propagation.end()) {
-        propagation.erase(existing);
+    if (existing == candidates.end() || *existing < info) {
+      if (existing != candidates.end()) {
+        candidates.erase(existing);
       }
-      auto pos = std::upper_bound(propagation.begin(), propagation.end(), info);
-      propagation.insert(pos, info);
+      auto pos = std::upper_bound(candidates.begin(), candidates.end(), info);
+      candidates.insert(pos, info);
+    }
+  };
+
+  auto allowPasC = [this](TensorView* from, TensorView* to) {
+    if (selector_ == nullptr) {
+      return true;
     }
+    return selector_->allowPasC(from, to);
   };
 
-  while (!propagation.empty()) {
-    auto next_hop = propagation.back();
-    propagation.pop_back();
+  auto allowCasP = [this](TensorView* from, TensorView* to) {
```

Comment:
Really interesting that you split these up.

Reply:
This has a real use case in my computeAt PR. In https://github.com/csarofeen/pytorch/pull/1743, I have
```C++
bool ConsumeAt::allowPasC(TensorView* from, TensorView* to) {
  return consume_.count(to) > 0;
}

bool ConsumeAt::allowCasP(TensorView* from, TensorView* to) {
  // If the producer is in the consume set, then the consumer must also be
  // replayed to obtain a compatible loop structure so that this producer
  // can be consumed in this loop.
  return consume_.count(from) > 0 || consume_.count(to) > 0;
}
```